x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxRequest.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.request.ibmwatsonx;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.common.socket.SocketAccess;
+import org.elasticsearch.xpack.inference.external.request.Request;
+import org.elasticsearch.xpack.inference.services.settings.DefaultSecretSettings;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+public interface IbmWatsonxRequest extends Request {
+
+    static void decorateWithBearerToken(HttpPost httpPost, DefaultSecretSettings secretSettings) {
+        final Logger logger = LogManager.getLogger(IbmWatsonxRequest.class);
+        String bearerTokenGenUrl = "https://iam.cloud.ibm.com/identity/token";
+        String bearerToken = "";
+
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            HttpPost httpPostForBearerToken = new HttpPost(bearerTokenGenUrl);
+
+            String body = "grant_type=urn:ibm:params:oauth:grant-type:apikey&apikey=" + secretSettings.apiKey().toString();
+            ByteArrayEntity byteEntity = new ByteArrayEntity(body.getBytes(StandardCharsets.UTF_8));
+
+            httpPostForBearerToken.setEntity(byteEntity);
+            httpPostForBearerToken.setHeader(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+            bearerToken = SocketAccess.doPrivileged(() -> {
+                HttpResponse response = httpClient.execute(httpPostForBearerToken);
+                HttpEntity entity = response.getEntity();
+                Map<String, Object> map;
+                try (InputStream content = entity.getContent()) {
+                    XContentType xContentType = XContentType.fromMediaType(entity.getContentType().getValue());
+                    map = XContentHelper.convertToMap(xContentType.xContent(), content, false);
+                }
+                return (String) map.get("access_token");
+            });
+        } catch (Exception e) {
+            logger.error("Failed to add Bearer token to the request", e);
```

Comment:
I changed this part to omit logging the exception since the exception may contain access token.

@prwhelan What is the best way to plumb the error back to the user? Should we throw another exception from the catch block?

Reply:
> I assume if we send an empty bearer token, this will ultimately fail the request with a 401 or 403 from IBM, whereas the actual error comes from this call and could either be a 500 if we made a mistake or another 401/403 if the API Key was wrong or something.

If the bearer token is empty, the user will receive 401 authentication error

Reply:
> @prwhelan What is the best way to plumb the error back to the user? Should we throw another exception from the catch block?

Yeah that would be fine, an exception thrown here will be caught in https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java#L127

We may want to reuse the response handler to make the distinction between a retryable exception and non-retryable, but that can also be a later change

Reply:
I updated the code to throw an exception now.